File: eurostag-indexes/src/TimeSeries.cpp
// Copyright (c) 2016, All partners of the iTesla project (http://www.itesla-project.eu/consortium)
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

/**
 * @file TimeSeries.cpp
 * @author Geoffroy Jamgotchian <geoffroy.jamgotchian at rte-france.com>
 */

#include <stdexcept>
#include "TimeSeries.h"

namespace itesla {

    TimeSerieType toTimeSerieType(const std::string& txt) {
        if (txt == "FREQUENCY") {
            return TimeSerieType::FREQUENCY;
        } else if (txt == "ANGLE") {
            return TimeSerieType::ANGLE;
        } else if (txt == "UNDER_VOLTAGE_AUTOMATON") {
            return TimeSerieType::UNDER_VOLTAGE_AUTOMATON;
        } else if (txt == "OVER_VOLTAGE_AUTOMATON") {
            return TimeSerieType::OVER_VOLTAGE_AUTOMATON;
        } else if (txt == "UNDER_SPEED_AUTOMATON") {
            return TimeSerieType::UNDER_SPEED_AUTOMATON;
        } else if (txt == "OVER_SPEED_AUTOMATON") {
            return TimeSerieType::OVER_SPEED_AUTOMATON;
        }  else if (txt == "INTEGRATION_STEP") {
            return TimeSerieType::INTEGRATION_STEP;
        }  else if (txt == "GENERATOR_ACTIVE_POWER") {
            return TimeSerieType::GENERATOR_ACTIVE_POWER;
        } else {
            throw std::runtime_error("Invalid time serie type " + txt);
        }
    }

    std::string toStr(TimeSerieType type) {
        switch (type) {
            case TimeSerieType::FREQUENCY:
                return "FREQUENCY";

            case TimeSerieType::ANGLE:
                return "ANGLE";

            case TimeSerieType::UNDER_VOLTAGE_AUTOMATON:
                return "UNDER_VOLTAGE_AUTOMATON";

            case TimeSerieType::OVER_VOLTAGE_AUTOMATON:
                return "OVER_VOLTAGE_AUTOMATON";

            case TimeSerieType::UNDER_SPEED_AUTOMATON:
                return "UNDER_SPEED_AUTOMATON";

            case TimeSerieType::OVER_SPEED_AUTOMATON:
                return "OVER_SPEED_AUTOMATON";

            case TimeSerieType::INTEGRATION_STEP:
                return "INTEGRATION_STEP";

            case TimeSerieType::GENERATOR_ACTIVE_POWER:
                return "GENERATOR_ACTIVE_POWER";

            case TimeSerieType::UNKNOWN:
                return "UNKNOWN";

            default: throw std::runtime_error("TODO");
        }
    }
}

File: math/src/main/java/com/powsybl/math/timeseries/TimeSeries.java
/**
 * Copyright (c) 2017, RTE (http://www.rte-france.com)
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
package com.powsybl.math.timeseries;

import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.core.JsonToken;
import com.powsybl.commons.json.JsonUtil;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.io.Writer;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Objects;
import java.util.stream.Stream;

/**
 * @author Geoffroy Jamgotchian <geoffroy.jamgotchian at rte-france.com>
 */
public interface TimeSeries<P extends AbstractPoint, C extends ArrayChunk<P>> extends Iterable<P> {

    TimeSeriesMetadata getMetadata();

    Stream<P> stream();

    Iterator<P> iterator();

    void writeJson(JsonGenerator generator);

    static void writeJson(JsonGenerator generator, List<? extends TimeSeries> timeSeriesList) {
        Objects.requireNonNull(timeSeriesList);
        for (TimeSeries timeSeries : timeSeriesList) {
            timeSeries.writeJson(generator);
        }
    }

    static void writeJson(Writer writer, List<? extends TimeSeries> timeSeriesList) {
        JsonUtil.writeJson(writer, generator -> writeJson(generator, timeSeriesList));
    }

    static void writeJson(Path file, List<? extends TimeSeries> timeSeriesList) {
        JsonUtil.writeJson(file, generator -> writeJson(generator, timeSeriesList));
    }

    static void parseChunks(JsonParser parser, TimeSeriesMetadata metadata, List<TimeSeries> timeSeriesList) {
        Objects.requireNonNull(metadata);
        List<DoubleArrayChunk> doubleChunks = new ArrayList<>();
        List<StringArrayChunk> stringChunks = new ArrayList<>();
        ArrayChunk.parseJson(parser, metadata.getDataType(), doubleChunks, stringChunks);
        if (metadata.getDataType() == TimeSeriesDataType.DOUBLE) {
            timeSeriesList.add(new StoredDoubleTimeSeries(metadata, doubleChunks));
        } else if (metadata.getDataType() == TimeSeriesDataType.STRING) {
            timeSeriesList.add(new StringTimeSeries(metadata, stringChunks));
        } else {
            throw new AssertionError("Unexpected time series data type " + metadata.getDataType());
        }
    }

    static List<TimeSeries> parseJson(JsonParser parser) {
        Objects.requireNonNull(parser);
        List<TimeSeries> timeSeriesList = new ArrayList<>();
        try {
            TimeSeriesMetadata metadata = null;
            JsonToken token;
            while ((token = parser.nextToken()) != null) {
                if (token == JsonToken.FIELD_NAME) {
                    String fieldName = parser.getCurrentName();
                    switch (fieldName) {
                        case "metadata":
                            metadata = TimeSeriesMetadata.parseJson(parser);
                            break;
                        case "chunks":
                            if (metadata == null) {
                                throw new AssertionError("metadata is null");
                            }
                            parseChunks(parser, metadata, timeSeriesList);
                            metadata = null;
                            break;
                        default:
                            break;
                    }

                }
            }
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
        return timeSeriesList;
    }

    static List<TimeSeries> parseJson(String json) {
        return JsonUtil.parseJson(json, TimeSeries::parseJson);
    }

    static List<TimeSeries> parseJson(Path file) {
        return JsonUtil.parseJson(file, TimeSeries::parseJson);
    }
}
